lib/pure/strutils.nim

```diff
@@ -2264,19 +2264,34 @@ proc insertSep*(s: string, sep = '_', digits = 3): string {.noSideEffect,
   ## if `s` contains a number.
   runnableExamples:
     doAssert insertSep("1000000") == "1_000_000"
-
-  var L = (s.len-1) div digits + s.len
+  var parts, prefix: string
+  let hasPrefix = isDigit(s[s.low]) == false
+  if hasPrefix:
+    var idx:int
+    prefix.add s[s.low]
+    for i in (s.low + 1)..s.high:
+      idx = i
+      if not isDigit(s[i]):
+        prefix.add s[i]
+      else:
+        break
+    parts = s.substr(idx)
+  else:
+    parts.shallowCopy s
```

Comment:
Please avoid `shallowCopy`, it's semantic is confusing and ARC will feature move semantics which is superior instead.